### PR TITLE
EEG-BIDS pipeline - Derivatives support: annotations

### PR DESCRIPTION
**Pull request by laemtl**
This PR adds support for derivatives, specifically annotations.
Annotations are added using 2 files, *_annotations.json (metadata) and _annotations.tsv (data), both mandatory.
For a template:  ([tsv](https://github.com/bids-standard/bids-examples/blob/9893043f52ab1cec144d1b6069497ef48edc9570/eeg_face13/derivatives/BIDS-Lossless-EEG/sub-001/eeg/sub-001_task-faceFO_annotations.tsv)) ([json](https://github.com/bids-standard/bids-examples/blob/9893043f52ab1cec144d1b6069497ef48edc9570/eeg_face13/derivatives/BIDS-Lossless-EEG/sub-001/eeg/sub-001_task-faceFO_annotations.json)) 

To test 
The PR depends on an SQL patch from https://github.com/aces/Loris/pull/7345 

PyBIDS won't detect *_annotations.json and _annotations.tsv with validate turned on without the [force_index](https://bids-standard.github.io/pybids/generated/bids.layout.BIDSLayout.html) option, as it ignores files that do not follow a naming convention allowed by the specs. The annotation specs are currently discussed in[ BIDS Extension Proposal 21](https://docs.google.com/document/d/1PmcVs7vg7Th-cGC-UrX8rAhKUHIzOI-uIOh69_mvdlw/edit#heading=h.mqkmyp254xh6).

Related to #450

### CODE REVIEW

**Comment by cmadjar** [comment]
@laemtl I am looking through the LORIS-MRI issues and noticed that there is the following issue which is related to this PR I think: #450

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:606]: move this in a file `lib/database_lib/physiologicalannotationfile.py` where all queries about `physiological_annotation_file.py` would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:613]: move this in a file `lib/database_lib/physiologicalannotationparameter.py` where all queries to `physiological_annotation_parameter` would go

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:621]: move the select into a file `lib/database_lib/physiologicalannotationlabel.py` where all queries to `physiological_annotation_label` would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:629]: move the insert into the file `lib/database_lib/physiologicalannotationlabel.py` with the select statement above.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:636]: this particular case has not yet been refactored and can be tacked on later on. I just think that for new queries, we should try to put all the database queries in the /database_lib/ folder, one file per table queried. One less thing to refactor after ;)

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:659]: move this in a file `lib/database_lib/physiologicalannotationfile.py` where all queries about `physiological_annotation_file.py` would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:688]: move this in a file `lib/database_lib/physiologicalannotationlabel.py` where all queries about physiological_annotation_label.py would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:696]: move this in a file lib/database_lib/physiologicalannotationlabel.py where all queries about physiological_annotation_label.py would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:704]: move this in a file lib/database_lib/physiologicalannotationparameter.py where all queries about physiological_annotation_parameter.py would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:725]: move this in a file lib/database_lib/physiologicalannotationinstance.py where all queries about physiological_annotation_instance.py would go.

File: python/lib/physiological.py
Comment (cmadjar) [inline python/lib/physiological.py:732]: this one will stay here for now as this was not yet refactored.

File: python/lib/eeg.py
Comment (cmadjar) [inline python/lib/eeg.py:307]: if annotation are a type of derived data, shouldn't the handling of annotations be done under the section of `registere_derivatives_data` instead of being in the `register_raw_data` function?

File: python/lib/eeg.py
Comment (cmadjar) [inline python/lib/eeg.py:760]: the select statement should be in a function in `lib/database_lib/physiologicalannotationfile.py` file which will return the annotation_path. To keep all the database calls into the `database_lib` folder.

File: python/lib/eeg.py
Comment (cmadjar) [inline python/lib/eeg.py:956]: the query should go in a function in `lib/database_lib/physiologicalannotationarchive.py` file where all queries to the physiological_annotation_archive table would be.

File: python/lib/eeg.py
Comment (cmadjar) [inline python/lib/eeg.py:996]: the insert statement should be in a function in `lib/database_lib/physiologicalannotationarchive.py` file

**Review by cmadjar — CHANGES_REQUESTED** [review]
@laemtl I did a review of your PR but did not test it yet. All my comments are related to a refactoring we decided to do on the Perl side and then to apply on the Python side as well.

Basically, all calls (select, insert, update...) to the database should be done in individual files names as the table queried in the `lib/database_lib/<tablename>.py` to centralize all the calls to the DB in one place instead of all over the code. You can see some examples in the `lib/database_lib` folder already. The old code has not yet been refactored to reflect that but all new code should follow that design so we have less to refactor when we do tackle the refactoring of existing code more seriously. 

Most of the `bids_import` code has not yet been refactored but some newer code are following that new design.

Would it be OK for you to move the new database calls present in this PR to respective files in the `database_lib` folders?

Finally, since there are new tables for the annotation, it might be good to update the documentation of the physiological_* tables in https://github.com/aces/Loris-MRI/blob/main/docs/03-TechnicalInfrastructure.md. Maybe there are some other parts that need to be updated in the rest of the docs as well. https://github.com/aces/Loris-MRI/blob/main/docs

File: python/lib/bidsreader.py
Comment (cmadjar) [inline python/lib/bidsreader.py:87]:
```suggestion
        bids_layout = BIDSLayout(root=self.bids_dir, config=bids_config, ignore=exclude_arr, derivatives=True)
```

@laemtl see my comment on https://github.com/aces/Loris-MRI/issues/609#issuecomment-818237774

No need for the `force_index` option in the end.

**Comment by cmadjar** [comment]
@laemtl I added the Needs work before merging label to this one as well. When ready for another review, remove the label and tag me. Thank you :)

File: python/lib/eeg.py
Comment (laemtl) [reply python/lib/eeg.py:307]: The problem here is that register_derivatives_data only processes derivatives eeg. It depends on how we decide to store annotations, but most likely annotations will be embedded in the source folder and target a raw eeg file, which will not fit the case covered by register_derivatives_data.

File: python/lib/bidsreader.py
Comment (laemtl) [reply python/lib/bidsreader.py]: See my comment in #609. Your solution works if annotations are embedded in a derivatives folder only. It might be fine if we go this route. We need to discuss where annotations files should be stored for convenience beforehand.

Comment (cmadjar) [reply python/lib/bidsreader.py]: @laemtl should we discuss this tomorrow at the imaging call?

**Comment by cmadjar** [comment]
Added the label needs work before merging because of the conflicts with main

**Comment by laemtl** [comment]
@cmadjar I realized I forgot to push the last fixes on this one. Ready :)

**Review by cmadjar — CHANGES_REQUESTED** [review]
@laemtl was there a patch for `physiological_annotation_rel` sent to LORIS?

**Comment by laemtl** [comment]
> @laemtl was there a patch for `physiological_annotation_rel` sent to LORIS?

Yes, here: https://github.com/aces/Loris/pull/7551

**Comment by cmadjar** [comment]
Blocked by aces/Loris#7551 on the LORIS side.

**Review by cmadjar — APPROVED** [review]
Works great. Will have to wait for the merge of the SQL patch on the LORIS side before merging this PR though. I tagged Dave and added the `critical to release` label to the LORIS PR.

**Pull request merged**
